From aad97c18c7580d9b29e32026c6b6293225ca3932 Mon Sep 17 00:00:00 2001
From: Robert Wiegand <qa2113@motorola.com>
Date: Wed, 8 Oct 2014 12:20:21 -0500
Subject: [PATCH 085/959] IKLOCSEN-1725 drivers: misc: stml0xx: Add headset
 detect enable control

Add a new device tree setting for sensor hub headset
detect enable/disable.

Change-Id: I8157919b60a0c42050f3f1c754b4d7f5e5e9e959
Signed-off-by: Robert Wiegand <qa2113@motorola.com>
Reviewed-on: http://gerrit.mot.com/677325
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Jagatkumar Shah <jagat@motorola.com>
Reviewed-by: Vladimir Karpovich <vkarpovich@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/misc/stml0xx/stml0xx_core.c  |  5 ++++-
 drivers/misc/stml0xx/stml0xx_reset.c | 10 ++++++++++
 include/linux/stml0xx.h              |  1 +
 3 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/drivers/misc/stml0xx/stml0xx_core.c b/drivers/misc/stml0xx/stml0xx_core.c
index 0b67d7c2635..0ca8c22f669 100644
--- a/drivers/misc/stml0xx/stml0xx_core.c
+++ b/drivers/misc/stml0xx/stml0xx_core.c
@@ -407,14 +407,17 @@ static struct stml0xx_platform_data *stml0xx_of_init(struct spi_device *spi)
 	of_property_read_u32(np, "ct406_pulse_count",
 			     &pdata->ct406_pulse_count);
 
+	pdata->headset_detect_enable = 0;
 	pdata->headset_insertion_debounce = 0x01F4;
-	pdata->headset_removal_debounce = 0x001E;
+	pdata->headset_removal_debounce = 0x01F4;
 	pdata->headset_button_down_debounce = 0x0032;
 	pdata->headset_button_up_debounce = 0x0032;
 	pdata->headset_button_0_1_threshold = 0x0096;
 	pdata->headset_button_1_2_threshold = 0x012C;
 	pdata->headset_button_2_3_threshold = 0x01C2;
 	pdata->headset_button_3_upper_threshold = 0x02EE;
+	of_property_read_u32(np, "headset_detect_enable",
+			     &pdata->headset_detect_enable);
 	of_property_read_u32(np, "headset_insertion_debounce",
 			     &pdata->headset_insertion_debounce);
 	of_property_read_u32(np, "headset_removal_debounce",
diff --git a/drivers/misc/stml0xx/stml0xx_reset.c b/drivers/misc/stml0xx/stml0xx_reset.c
index acfdc25de68..04211282868 100644
--- a/drivers/misc/stml0xx/stml0xx_reset.c
+++ b/drivers/misc/stml0xx/stml0xx_reset.c
@@ -182,7 +182,17 @@ void stml0xx_initialize_work_func(struct work_struct *work)
 		dev_err(&ps_stml0xx->spi->dev,
 			"unable to write headset settings %d", err);
 		ret_err = err;
+	} else {
+		buf[0] = pdata->headset_detect_enable & 0xFF;
+		err = stml0xx_spi_send_write_reg_reset(HEADSET_CONTROL, buf,
+			1, RESET_NOT_ALLOWED);
+		if (err < 0) {
+			dev_err(&ps_stml0xx->spi->dev,
+				"Unable to wrie headset detect enable");
+			ret_err = err;
+		}
 	}
+
 	err = stml0xx_led_set_reset(&ps_stml0xx->led_cdev,
 			RESET_NOT_ALLOWED);
 	if (err < 0)
diff --git a/include/linux/stml0xx.h b/include/linux/stml0xx.h
index eeb831796d8..d24844f664f 100644
--- a/include/linux/stml0xx.h
+++ b/include/linux/stml0xx.h
@@ -540,6 +540,7 @@ struct stml0xx_platform_data {
 	int ct406_undetect_threshold;
 	int ct406_recalibrate_threshold;
 	int ct406_pulse_count;
+	int headset_detect_enable;
 	int headset_insertion_debounce;
 	int headset_removal_debounce;
 	int headset_button_down_debounce;
-- 
2.11.0

